### PR TITLE
Texture3d support

### PR DESCRIPTION
**Pull request by artur-trzesiok**
Working example from K3D-jupyter:
- https://www.youtube.com/watch?v=F87SoSF9XAA
- https://www.youtube.com/watch?v=s3nU9fkT5L0

### CODE REVIEW

**Comment by moraxy** [comment]
`textures/Texture3D.js` seems to be missing - I'm getting a build error:
```
$ npm run-script build

> three@0.94.0 build /tmp/tmp.jm7GRQ5b8Z/three.js-mirror
> rollup -c


src/Three.js → build/three.js, build/three.module.js...
[!] Error: Could not resolve './textures/Texture3D.js' from src/Three.js
Error: Could not resolve './textures/Texture3D.js' from src/Three.js
    at error (/tmp/tmp.jm7GRQ5b8Z/three.js-mirror/node_modules/rollup/dist/rollup.js:199:15)
    at /tmp/tmp.jm7GRQ5b8Z/three.js-mirror/node_modules/rollup/dist/rollup.js:20784:25
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
    at Function.Module.runMain (module.js:695:11)
    at startup (bootstrap_node.js:191:16)
    at bootstrap_node.js:612:3

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! three@0.94.0 build: `rollup -c`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the three@0.94.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     ~/.npm/_logs/2018-07-23T14_47_03_378Z-debug.log
```

**Comment by takahirox** [comment]
Nice work. Can you add an 3D texture example under `examples/`?

**Comment by artur-trzesiok** [comment]
Probably. I don't have time right now to do this. I will try during a weekend.

**Comment by Mugen87** [comment]
Adding a new doc page would also be great. BTW: What would happen if the user uses `Texture3D` with a WebGL 1 context? It's maybe a good idea to detect this and provide an informative warning.

**Comment by takahirox** [comment]
I'm thinking how `TextureLoader` supports 3D texture. I'm not familiar with 3D texture yet. Can we detect if image is for 3D texture from the filename extension or something?

File: src/renderers/webgl/WebGLUtils.js [hunk 1]
```diff
@@ -51,6 +51,7 @@ function WebGLUtils( gl, extensions ) {
 		if ( p === LuminanceAlphaFormat ) return gl.LUMINANCE_ALPHA;
 		if ( p === DepthFormat ) return gl.DEPTH_COMPONENT;
 		if ( p === DepthStencilFormat ) return gl.DEPTH_STENCIL;
+		if ( p === RedFormat ) return gl.RED;
```
Comment (takahirox) [inline src/renderers/webgl/WebGLUtils.js]: `gl.RED` is available only in WebGL2 so I think we need to check if `gl` is `WebGL2RenderingContext`.

    if ( capabilities.isWebGL2 && p === RedFormat ) return gl.RED;

File: src/renderers/webgl/WebGLUniforms.js [hunk 2]
```diff
@@ -53,6 +53,7 @@ import { CubeTexture } from '../../textures/CubeTexture.js';
 import { Texture } from '../../textures/Texture.js';
 
 var emptyTexture = new Texture();
+var emptyTexture3d = new Texture();
```
Comment (takahirox) [inline src/renderers/webgl/WebGLUniforms.js]: Isn't here `new Texture3D()`?

File: src/textures/Texture3D.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+ */
+
+import { Texture } from './Texture.js';
+import { NearestFilter } from '../constants.js';
+
+function Texture3D( data, width, height, length, format, type, mapping, wrapS, wrapT, magFilter, minFilter, anisotropy, encoding ) {
+
+	Texture.call( this, null, mapping, wrapS, wrapT, magFilter, minFilter, format, type, anisotropy, encoding );
+
+	this.image = { data: data, width: width, height: height, length: length };
```
Comment (takahirox) [inline src/textures/Texture3D.js]: Is there any reasons why you're targeting only data texture, not image texture?

**Comment by DavidPeicho** [comment]
I am still not convinced about the automatic deduction of the internal format. In my case for instance, I use particular texture with internal format, such as RG8UI. Maybe we could find a good way to add this internal format somewhere. What do you think? @takahirox

File: src/textures/Texture3D.js (hunk 3 above)
Comment (artur-trzesiok) [reply src/textures/Texture3D.js]: Very simple reason - i only need a data texture. I the other hand - writing support for 3d images dataformat is a separate task from my point of view.

**Comment by artur-trzesiok** [comment]
New examples:
k3d-1532603480276
k3d-1532603595150

**Comment by gkjohnson** [comment]
This is awesome! Are there common standardized formats for 3d textures?

File: src/textures/Texture3D.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (gkjohnson) [reply src/textures/Texture3D.js]: fwiw Unity calls this dimension `depth` on their Texture3D

Comment (takahirox) [reply src/textures/Texture3D.js]: OK. Then I think `DataTexture3D` would sound better rather than `Texture3D` in this PR.

Comment (gkjohnson) [reply src/textures/Texture3D.js]: Would there be a significant difference between a DataTexture3D and just a Texture3D in this case? The width, height, and depth need to be separately specified even if using an image to bind data:
https://developer.mozilla.org/en-US/docs/Web/API/WebGL2RenderingContext/texImage3D

As a little bit of an aside, how is the data interpreted when binding a 2d image as a texture 3d?

**Comment by takahirox** [comment]
Can you add an example under `examples/`? https://github.com/mrdoob/three.js/tree/dev/examples Users and developers can see 3D texture works on Three.js.

**Comment by takahirox** [comment]
I locally made some examples
- https://twitter.com/superhoge/status/1034875111923482625
- https://twitter.com/superhoge/status/1034906630079631365
- https://twitter.com/superhoge/status/1035252229073186817

and confirmed this PR seems working.

@artur-trzesiok seems to have been busy. So I wanna suggest to merge this PR so far. And we do renaming, adding an example, brushing up, and so on in other PRs.

**Comment by takahirox** [comment]
Currently no 3D texture(`sampler3D` and `texture( sampler3D, vec3 )`) in default shaders. So users must use `[Raw]ShaderMaterial` for 3D texture. I think it's good start. I heard major use cases of 3D texture are Volume rendering and sometimes gpgpu. In general users write their shader codes for them.

This leads to a question to me. (I'm studying but still new to 3D texture.) Are there any other use cases and/or do we need to have material or something which can handle 3D texture?

**Comment by mrdoob** [comment]
> @artur-trzesiok seems to have been busy. So I wanna suggest to merge this PR so far. And we do renaming, adding an example, writing documents, brushing up, and so on in other PRs.

Sounds good!

**Comment by mrdoob** [comment]
Thanks!

File: src/renderers/WebGLRenderer.js [hunk 4]
```diff
@@ -2353,6 +2353,17 @@ function WebGLRenderer( parameters ) {
 
 	}() );
 
+	this.setTexture3D = ( function () {
```
Comment (mrdoob) [inline src/renderers/WebGLRenderer.js]: @takahirox we can probably remove this as this function?

File: src/renderers/webgl/WebGLUniforms.js (hunk 2 above)
Comment (mrdoob) [reply src/renderers/webgl/WebGLUniforms.js]: Probably 👍

**Comment by takahirox** [comment]
Cool. Do you also want to make PR to Three.js?

**Comment by almarklein** [comment]
Sure, I can do that.

**Comment by takahirox** [comment]
Wondering if we can have two 3D Texture examples. 

One is @almarklein 's example. Users can see how nice 3D Texture + Volume rendering is. And users could reuse volume rendering shaders.

Another one is based on my simple and light example https://twitter.com/superhoge/status/1034875111923482625 Users can know how to use 3D Texture in Three.js.

**Comment by mrdoob** [comment]
Sounds good!

**Pull request merged**